dnf5: Command "do"

jrohel opened the pull request:
Closes: https://github.com/rpm-software-management/dnf5/issues/1223

Required modification of CI tests: https://github.com/rpm-software-management/ci-dnf-stack/pull/1677

Example - Installing "pkg1" and "pkg2" from repositories whose repoid matches "my_repo*", installing "pkg3" from the "fedora" repository, and upgrading "pkg4" from the "my_repo_updates" repository. The required dependencies can be taken from all enabled repositories:
`dnf do --action=install --use-repo='my_rep*' pkg1 pkg2 --use-repo=fedora pkg3 --action=upgrade --use-repo=my_repo_updates pkg4`

dcantrell reviewed (approved):
Looks reasonable to me.  No comments.

ppisar commented:
Do you foresee any other options related to repositories than an option for binding a package to a repository? If you don't, I would propose renaming the `--use-repo` option to `--repo`. Or is the clash with a global `--repo` option a problem?

jrohel commented:
@ppisar 
> Or is the clash with a global --repo option a problem?

Yes. The existing `--repo` has a different meaning. It defines a list of allowed repositories.

Example:
`dnf --repo=fedora,updates do --use-repo=fedora --action=install pkg`
This means installing `pkg` from fedora repository, but dependencies can be taken from `fedora` and `updates`.

In addition, we allow the use of global options even after the command. So the above example can also be written as:
`dnf do --repo=fedora,updates --use-repo=fedora --action=install pkg`
Changing the meaning of `--repo` after the command could confuse someone. That's why I prefer the other name `--use-repo`.

m-blaha commented:
Also the new command needs documentation and tests. Do you plan to work on CI tests for the `do` command?

Conan-Kudo commented:
I think instead of `do` the verb `run` makes more sense.

My thinking is that it makes sense to be able to read it like a command sentence: "DNF: run action <action> with <options>". It's easier to reason that way too. It also makes sense if in a DNF shell environment if/when that gets ported to DNF5. Then you can just replace "run" with a shell prompt.

jrohel commented on dnf5/commands/do/do.cpp:
```diff
@@ -0,0 +1,347 @@
+        advisory_newpackage->get_value(),
+        advisory_severity->get_value(),
+        advisory_bz->get_value(),
+        advisory_cve->get_value());
```
Thanks. I will rebase and update code.

jrohel commented on dnf5/commands/do/do.cpp:
```diff
@@ -0,0 +1,347 @@
+    {
+        auto item_use_repo_opt = parser.add_new_named_arg("use-repo");
+        item_use_repo_opt->set_long_name("use-repo");
+        item_use_repo_opt->set_description(_("Use specified repositories. Dependencies from all enabled."));
```
I tried to make the help as short as possible, but still understandable. But OK, I'll use the suggested longer version.

jrohel commented on dnf5/commands/do/do.cpp:
```diff
@@ -0,0 +1,347 @@
+    }
+
+    for (auto & item : downgrade_items) {
+        // TODO(jrohel): Do we want to apply advisories filter on downgrade?
```
Thanks. I will remove code.

evan-goode commented:
> I think instead of `do` the verb `run` makes more sense.

I like `do` over `run` since `run` has sort of a "fork and exec" connotation, as if I'm giving DNF a script to run. `do` feels more like I'm telling DNF to do a list of tasks itself.

We decided to call the `--use-repo` flag `--from-repo` instead since `--use-repo` is a bit too nonspecific (aren't all configured repositories "used" by default?).

jrohel commented:
I rebased the code. And replaced `--use-repo` with `--from-repo`.

The failing CI tests seem to be unrelated to this PR.
Failing scenarios:
  dnf/option-hint.feature:4  command specific long option used with root dnf command
  dnf/option-hint.feature:37  multiple command specific long options used with root dnf command

kontura commented:
> Also the new command needs documentation and tests. Do you plan to work on CI tests for the `do` command?

We have agreed to deliver the docs and tests in subsequent PRs.

Tests pass locally.
Thank you.
